frontend/include/chpl/uast/Array.h

```diff
@@ -149,79 +149,65 @@ class Array final : public AstNode {
    */
   class FlatteningArrayIterator {
    private:
-    // Stack of current iterator positions, one for each dimension.
-    // The bottom iterates over the array itself, and the top iterates
-    // over a row of innermost dimension.
-    std::vector<AstList::const_iterator> rowIterStack;
-    // Iterator marking the end of the outermost dimension (the array itself)
-    AstList::const_iterator topLevelEnd;
+    // Stack of current row iterator positions, one for each dimension. The
+    // bottom iterates over the array itself, and the top iterates over a row of
+    // innermost dimension.
+    // Each entry is a pair of (current, end) iterators.
+    std::vector<std::pair<AstList::const_iterator, AstList::const_iterator>>
+        rowIterStack;
 
     /*
      * Descend to the innermost array dimension, adding an iterator for each
      * dimension along the way.
      */
     void descendDims() {
       CHPL_ASSERT(!rowIterStack.empty() && "should not be possible");
-      while (auto row = (*rowIterStack.back())->toArrayRow()) {
-        this->rowIterStack.push_back(row->begin());
+      while (auto row = (*rowIterStack.back().first)->toArrayRow()) {
+        this->rowIterStack.emplace_back(row->begin(), row->end());
       }
     }
 
    public:
     FlatteningArrayIterator(const Array* iterand, bool end = false) {
       if (end) {
-        rowIterStack.push_back(iterand->end());
+        rowIterStack.emplace_back(iterand->end(), iterand->end());
       } else {
-        rowIterStack.push_back(iterand->begin());
+        rowIterStack.emplace_back(iterand->begin(), iterand->end());
         descendDims();
       }
-      topLevelEnd = iterand->end();
     }
```

Comment:
Rather than a constructor with a boolean argument, consider two static builder functions. I think it's better stylistically. You might have:

```C++
 private:
  FlatteningArrayIterator(AstList::const_iterator begin, AstList::const_iterator end) {
    rowIterStack.emplace_back(begin, end);
  }

 public:
  static normal(const Array* iterand) {
    FlatteningArrayIterator ret(iterand->begin(), iterand->end());
    ret.descendDims();
    return ret;
  }

  static end(const Array* iterand) {
    return FlatteningArrayIterator(iterand->end(), iterand->end());
  }
```